### PR TITLE
Improve AI Turn times

### PR DESCRIPTION
**Pull request by Morlic-fo**
By turn 200 with a decently scaling AI, I get about 30% reduction in turn times.

Most improvements come from avoiding multiple computations by adequate caching of results.

### CODE REVIEW

File: default/python/AI/AIDependencies.py [hunk 1]
```diff
@@ -143,11 +143,11 @@
 BASE_PLANET_STEALTH = 5
 
 STEALTH_STRENGTHS_BY_SPECIES_TAG = {
-    "BAD_STEALTH": -PRIMARY_FOCS_STEALTH_LEVELS["LOW_STEALTH"],
-    "GOOD_STEALTH": PRIMARY_FOCS_STEALTH_LEVELS["LOW_STEALTH"],
-    "GREAT_STEALTH": PRIMARY_FOCS_STEALTH_LEVELS["MEDIUM_STEALTH"],
-    "ULTIMATE_STEALTH": PRIMARY_FOCS_STEALTH_LEVELS["HIGH_STEALTH"],
-    "INFINITE_STEALTH": 500,  # used by super testers
+    "BAD": -PRIMARY_FOCS_STEALTH_LEVELS["LOW_STEALTH"],
```
Comment (Cjkjvfnby) [inline default/python/AI/AIDependencies.py:158]: Does this speedup AI? :)

File: default/python/AI/ColonisationAI.py [hunk 2]
```diff
@@ -190,24 +195,21 @@ def galaxy_is_sparse():
              ((avg_empire_systems >= 35) and (setup_data.shape != fo.galaxyShape.elliptical))))
 
 
-def rate_piloting_tag(tag_list):
+def rate_piloting_tag(species_name):
```
Comment (Cjkjvfnby) [inline default/python/AI/ColonisationAI.py:201]: Does it change during gameplay? Can we cache it forever with decorator?

File: default/python/AI/ColonisationAI.py [hunk 3]
```diff
@@ -453,6 +455,7 @@ def get_colony_fleets():
 # TODO: clean up suppliable versus annexable
 def assign_colonisation_values(planet_ids, mission_type, species, detail=None, return_all=False):
     """Creates a dictionary that takes planetIDs as key and their colonisation score as value."""
+    empire_research_list = [element.tech for element in fo.getEmpire().researchQueue]
```
Comment (Cjkjvfnby) [inline default/python/AI/ColonisationAI.py]: can it be tuple? For immutability reasons.

File: default/python/AI/ColonisationAI.py [hunk 4]
```diff
@@ -651,13 +660,14 @@ def evaluate_planet(planet_id, mission_type, spec_name, detail=None):
     planet_supply += sum(AIDependencies.building_supply[bld_type].get(int(psize), 0)
                          for psize in [-1, planet.size] for bld_type in bld_types)
     planet_supply += sum(AIDependencies.SUPPLY_MOD_SPECIALS[_special].get(int(psize), 0)
-                         for psize in [-1, planet.size] for _special in
-                         set(planet.specials).union(system.specials).intersection(AIDependencies.SUPPLY_MOD_SPECIALS))
+                         for _special in set(planet.specials).union(system.specials).intersection(AIDependencies.SUPPLY_MOD_SPECIALS)
```
Comment (Cjkjvfnby) [inline default/python/AI/ColonisationAI.py]: You can extract it to the variable for readability
`set(planet.specials).union(system.specials).intersection(AIDependencies.SUPPLY_MOD_SPECIALS)`

File: default/python/AI/InvasionAI.py [hunk 5]
```diff
@@ -173,7 +173,7 @@ def get_invasion_fleets():
                 continue
             # TODO: have TroopShipDesigner give the expected number of troops including species effects directly
             troops_per_ship = best_base_trooper_here.troopCapacity
-            species_troop_grade = CombatRatingsAI.get_species_troops_grade(loc_planet.speciesName)
+            species_troop_grade = get_species_tag_grade(loc_planet.speciesName, 'ATTACKTROOPS')
```
Comment (Cjkjvfnby) [inline default/python/AI/InvasionAI.py]: `ATTACKTROOPS`  can we create a constant for tags?  I want to be sure that typo is not possible here.

File: default/python/AI/freeorion_tools/_freeorion_tools.py [hunk 6]
```diff
@@ -143,10 +144,12 @@ def chat_human(message):
     debug("Chat Message to human: %s", remove_tags(message))
 
 
-def cache_by_session(func):
+def cache_for_session(func):
```
Comment (Cjkjvfnby) [inline default/python/AI/freeorion_tools/_freeorion_tools.py:147]: Now when we got python3 we can just:
`cache_for_session = functools.lru_cache`

https://docs.python.org/3/library/functools.html?highlight=lru#functools.lru_cache

**Review by Cjkjvfnby — APPROVED** [review]
Looks great.  

All my comments are optional.

File: default/python/AI/AIDependencies.py (hunk 1 above)
Comment (Morlic-fo) [reply default/python/AI/AIDependencies.py:158]: This is a change to bring the dict into the form expected by the cached function get_species_tag_grade. So, indirectly, yes.

File: default/python/AI/freeorion_tools/_freeorion_tools.py (hunk 6 above)
Comment (Morlic-fo) [reply default/python/AI/freeorion_tools/_freeorion_tools.py:147]: Beyond the scope of this PR but probably good to keep in mind for the future.

**Pull request merged**